src/controllers/AlgorithmController.ts

```diff
@@ -0,0 +1,100 @@
+import { OptimizationAlgorithm } from '@/models/OptimizationAlgorithm';
+import winston = require('winston');
+import OptimizationExecutionModel, { OptimizationExecution } from '@/models/OptimizationExecution';
+import Ingredient from './IngredientInterface';
+import DockerController from './DockerController';
+import IntermediateResult from '@/models/IntermediateResult';
+import { promises as fs } from 'fs';
+import path from 'path';
+import config from '@/config.json';
+import { ResourceInstance } from '@/models/ResourceInstance';
+
+export default class AlgorithmController implements Ingredient {
+  // region public static methods
+  // endregion
+
+  // region private static methods
+  // endregion
+
+  // region public members
+  public optimizationAlgorithm: OptimizationAlgorithm;
+  // endregion
+
+  // region private members
+  private dockerController: DockerController;
+  private filePathForDataExchange!: string;
+  // endregion
+
+  // region constructor
+  constructor(optimizationAlgorithm: OptimizationAlgorithm) {
+    this.dockerController = new DockerController();
+    this.optimizationAlgorithm = optimizationAlgorithm;
+  }
+  // endregion
+
+  // region public methods
+  public async execute(input: IntermediateResult): Promise<OptimizationExecution> {
+    const executionInstance = new OptimizationExecutionModel();
+    executionInstance.optimizationAlgorithm = this.optimizationAlgorithm;
+    await this.createDirectoryForDataExchange(executionInstance.identifier);
+    this.writeRequiredAlgorithmInputFiles(input);
+
+    try {
+      await executionInstance.save();
+      const containerName = executionInstance.containerName;
+
+      this.dockerController.run(
+        this.optimizationAlgorithm.imageIdentifier,
+        ['/bin/bash', '-c', 'sleep 30s'],
+        process.stdout,
+        { name: containerName })
+        .then((container) => {
+          executionInstance.terminate(container.output.StatusCode);
+          winston.debug(`Container ${containerName} exited with code ${container.output.StatusCode}`);
+        });
+
+      return executionInstance;
+    } catch (error) {
+      return new Promise((resolve, reject) => {
```

Comment:
This way to reject Promises looks more complicated then it has to be. You can achieve the same result by simply throwing an error inside the async function. Related question: https://stackoverflow.com/questions/42453683/how-to-reject-in-async-await-syntax

Consider this comment for all places this code snipped is used.